### PR TITLE
Remove compile time algorithm checks from libssl

### LINKED ISSUE #13184: Don't offer DH ciphersuites if we don't have DH available
Currently we detect what key exchange algorithms we have in order to create the supported_group extension. If we don't have any DH algorithms that we won't add DH groups to the supported_group extension.However this isn't the case with ciphersuites. We will still offer DH ciphersuites even if the DH algorithm is not available in any provider.

As discussed in https://github.com/openssl/openssl/pull/13138#issuecomment-712243331

### LINKED ISSUE #13616: There should be no algorithm specific ifndef OPENSSL_NO_ALGORITHM in libssl
As crypto implementations are provided and the libssl should now call high level APIs from libcrypto only the algorithm specific ifndefs should not be present in libssl code at all.

This of course requires properly handling cases where an algorithm supported by libssl is not supported by a provider but that should be already part of the cleanup necessary for libssl support for external providers.

### PR DESCRIPTION
**Pull request by mattcaswell**
This PR removes (almost) all compile time checks for the availability of crypto algorithms from libssl and replaces them with runtime checks where necessary.

This is necessary because:

- Even if OpenSSL is compiled with full algorithm support, those algorithms may not be available at runtime if the relevant providers are not present
- If OpenSSL is compiled without full algorithm support, the missing algorithms can be added back later at runtime if providers that have those algorithms are loaded

Note that this does not mean the removal of all `OPENSSL_NO_XXX` guards from libssl. Only those relating to crypto algorithms. Therefore guards such as `OPENSSL_NO_NEXTPROTONEG`, `OPENSSL_NO_TLS1_3`, `OPENSSL_NO_CT` etc still exist.

There are a handful of algorithm specific guards that do still exist. These are exclusively in the tls_depr.c file and relate to deprecated APIs that use low level crypto objects.

One other note is that OPENSSL_NO_SRP still exists. SRP is not exposed via EVP at all and the low level API is not currently deprecated.

Finally, I have removed from Configure the cascade that "no-ec" and "no-dh" means "no-tls1_3". It is now perfectly possible that in a "no-ec" and "no-dh" build you could still use TLSv1.3 given the right providers and config. It does however mean that a *default* "no-ec" and "no-dh" build will fail to complete a handshake if both peers use TLSv1.3 without additional config steps.

This is marked as WIP because it includes all of the commits from #13139.

Fixes #13616
Fixes #13184

### CODE REVIEW

**Comment by mattcaswell** [comment]
I raised an issue relating to the SRP comment above in #13917

**Comment by mattcaswell** [comment]
This PR also fixes the remaining issue in #13767

**Comment by mattcaswell** [comment]
Rebased now that #13139 has been merged. Also taken out of WIP. Please take a look.

File: include/openssl/ssl.h.in [hunk 1]
```diff
@@ -194,14 +194,9 @@ extern "C" {
...
-#   define TLS_DEFAULT_CIPHERSUITES "TLS_AES_256_GCM_SHA384:" \
-                                    "TLS_CHACHA20_POLY1305_SHA256:" \
-                                    "TLS_AES_128_GCM_SHA256"
-#  else
-#   define TLS_DEFAULT_CIPHERSUITES "TLS_AES_256_GCM_SHA384:" \
+#  define TLS_DEFAULT_CIPHERSUITES "TLS_AES_256_GCM_SHA384:" \
+                                   "TLS_CHACHA20_POLY1305_SHA256:" \
                                    "TLS_AES_128_GCM_SHA256"
-#  endif
 # endif
```
Comment (beldmit) [inline include/openssl/ssl.h.in:200]: I have some doubts it will work.

For TLS versions <= 1.2 because there was some runtime detection of the ciphers necessary for a particular ciphersuite. I didn't find such a check in this code so it means a client will announce the TLS_CHACHA20_POLY1305_SHA256 ciphersuite even without any ChaCha support and server may choose it.

File: providers/common/capabilities.c [hunk 2]
```diff
@@ -135,23 +145,28 @@ static const OSSL_PARAM param_group_list[][10] = {
...
 #  ifndef FIPS_MODULE
     TLS_GROUP_ENTRY("brainpoolP256r1", "brainpoolP256r1", "EC", 25),
     TLS_GROUP_ENTRY("brainpoolP384r1", "brainpoolP384r1", "EC", 26),
     TLS_GROUP_ENTRY("brainpoolP512r1", "brainpoolP512r1", "EC", 27),
 #  endif
-    TLS_GROUP_ENTRY("x25519", "x25519", "X25519", 28),
-    TLS_GROUP_ENTRY("x448", "x448", "X448", 29),
+    TLS_GROUP_ENTRY("x25519", "X25519", "X25519", 28),
+    TLS_GROUP_ENTRY("x448", "X448", "X448", 29),
 # endif /* OPENSSL_NO_EC */
```
Comment (beldmit) [inline providers/common/capabilities.c:170]: I think the alias changes should be moved to a separate PR.

Comment (beldmit) [reply providers/common/capabilities.c:170]: I mean all the alias-related changes from this file.

File: include/openssl/ssl.h.in (hunk 1 above)
Comment (mattcaswell) [reply include/openssl/ssl.h.in:200]: > I didn't find such a check in this code

I didn't add a new check in this PR because one already exists in master:

https://github.com/openssl/openssl/blob/a2a5506b9329b978a2a5b11a518b9789446ad310/ssl/ssl_ciph.c#L327-L337

I just tested it, and can confirm that compiling this PR with "no-chacha" there are no CHACHA based ciphersuites offered by the client

Comment (beldmit) [reply include/openssl/ssl.h.in:200]: Got it. Thanks!

File: providers/common/capabilities.c (hunk 2 above)
Comment (mattcaswell) [reply providers/common/capabilities.c:170]: Without them things will break after this PR. You'll no longer be able to specify NIST names for groups, which you can prior to this PR. Therefore I think they should stay here.

**Comment by mattcaswell** [comment]
New commit pushed addressing a CI failure. Please take another look.

File: providers/common/capabilities.c [hunk 3]
```diff
@@ -135,23 +145,28 @@ static const OSSL_PARAM param_group_list[][10] = {
...
+    TLS_GROUP_ENTRY("P-521", "secp521r1", "EC", 24), /* Alias of above */
 #  ifndef FIPS_MODULE
     TLS_GROUP_ENTRY("brainpoolP256r1", "brainpoolP256r1", "EC", 25),
     TLS_GROUP_ENTRY("brainpoolP384r1", "brainpoolP384r1", "EC", 26),
     TLS_GROUP_ENTRY("brainpoolP512r1", "brainpoolP512r1", "EC", 27),
 #  endif
-    TLS_GROUP_ENTRY("x25519", "x25519", "X25519", 28),
-    TLS_GROUP_ENTRY("x448", "x448", "X448", 29),
+    TLS_GROUP_ENTRY("x25519", "X25519", "X25519", 28),
+    TLS_GROUP_ENTRY("x448", "X448", "X448", 29),
```
Comment (t8m) [inline providers/common/capabilities.c:169]: Why this particular change of case for x25519 and x448?

Comment (mattcaswell) [reply providers/common/capabilities.c:169]: This fixes a slight oddity in our x25519/x448 implementation revealed by this PR. In OpenSSL 1.1.1 the group name is "X25519" (i.e. uppercase X). We did not accept the lowercase form. So `openssl s_server -groups X25519` works in 1.1.1, but `openssl s_server -groups x25519` does not. However the IANA registry lists the official group name for TLS as `x25519` (i.e. lowercase x).

The reason we previously had uppercase X is because that's the way the internal OBJ database has it. But when we introduced the "TLS-GROUP" capability the first named parameter is supposed to be the IANA group name, and the 2nd parameter is supposed to be the name that we know it by internally.

Prior to this PR a configuration such as "-groups X25519" would first look up the name in the OBJ database, and if it didn't find it there, then it would check all the groups names from the "TLS-GROUP" capability. Hence both "openssl s_server -groups X25519" and "openssl s_server -groups x25519" both worked in master.

This PR simplifies that code to only run it off the "TLS-GROUP" capability. However it exposed this small discrepancy between the official name and the internal name. This fixes things so that the internal name is the same as how we actually know it internally, and means that "openssl s_server -X25519" continues to work.

Comment (t8m) [reply providers/common/capabilities.c:169]: Thanks for the explanation.

File: ssl/s3_lib.c [hunk 4]
```diff
@@ -3452,7 +3435,7 @@ long ssl3_ctrl(SSL *s, int cmd, long larg, void *parg)
     case SSL_CTRL_GET_FLAGS:
         ret = (int)(s->s3.flags);
         break;
-#if !defined(OPENSSL_NO_DH) && !defined(OPENSSL_NO_DEPRECATED_3_0)
+#if !defined(OPENSSL_NO_DEPRECATED_3_0)
```
Comment (t8m) [inline ssl/s3_lib.c:3438]: I wonder if this particular change makes sense as the CTRL requires the DH objects to be built in libcrypto to be functional.

File: ssl/s3_lib.c [hunk 5]
```diff
@@ -3477,7 +3460,7 @@ long ssl3_ctrl(SSL *s, int cmd, long larg, void *parg)
     case SSL_CTRL_SET_DH_AUTO:
         s->cert->dh_tmp_auto = larg;
         return 1;
-#if !defined(OPENSSL_NO_EC) && !defined(OPENSSL_NO_DEPRECATED_3_0)
+#if !defined(OPENSSL_NO_DEPRECATED_3_0)
```
Comment (t8m) [inline ssl/s3_lib.c:3463]: similarly as above the passed object is EC_KEY so it cannot work with NO_EC.

File: ssl/s3_lib.c [hunk 6]
```diff
@@ -3756,7 +3727,7 @@ long ssl3_callback_ctrl(SSL *s, int cmd, void (*fp) (void))
     int ret = 0;
 
     switch (cmd) {
-#if !defined(OPENSSL_NO_DH) && !defined(OPENSSL_NO_DEPRECATED_3_0)
+#if !defined(OPENSSL_NO_DEPRECATED_3_0)
```
Comment (t8m) [inline ssl/s3_lib.c:3730]: similarly to above comment

File: ssl/s3_lib.c [hunk 7]
```diff
@@ -3781,7 +3752,7 @@ long ssl3_callback_ctrl(SSL *s, int cmd, void (*fp) (void))
 long ssl3_ctx_ctrl(SSL_CTX *ctx, int cmd, long larg, void *parg)
 {
     switch (cmd) {
-#if !defined(OPENSSL_NO_DH) && !defined(OPENSSL_NO_DEPRECATED_3_0)
+#if !defined(OPENSSL_NO_DEPRECATED_3_0)
```
Comment (t8m) [inline ssl/s3_lib.c:3755]: and here

File: ssl/s3_lib.c [hunk 8]
```diff
@@ -3805,7 +3776,7 @@ long ssl3_ctx_ctrl(SSL_CTX *ctx, int cmd, long larg, void *parg)
     case SSL_CTRL_SET_DH_AUTO:
         ctx->cert->dh_tmp_auto = larg;
         return 1;
-#if !defined(OPENSSL_NO_EC) && !defined(OPENSSL_NO_DEPRECATED_3_0)
+#if !defined(OPENSSL_NO_DEPRECATED_3_0)
```
Comment (t8m) [inline ssl/s3_lib.c:3779]: and here

File: ssl/s3_lib.c [hunk 9]
```diff
@@ -4005,7 +3974,7 @@ long ssl3_ctx_ctrl(SSL_CTX *ctx, int cmd, long larg, void *parg)
 long ssl3_ctx_callback_ctrl(SSL_CTX *ctx, int cmd, void (*fp) (void))
 {
     switch (cmd) {
-#if !defined(OPENSSL_NO_DH) && !defined(OPENSSL_NO_DEPRECATED_3_0)
+#if !defined(OPENSSL_NO_DEPRECATED_3_0)
```
Comment (t8m) [inline ssl/s3_lib.c:3977]: and here as well

File: ssl/ssl_local.h [hunk 10]
```diff
@@ -2009,9 +2003,7 @@ typedef struct cert_st {
     CERT_PKEY *key;
 
     EVP_PKEY *dh_tmp;
-#ifndef OPENSSL_NO_DH
     DH *(*dh_tmp_cb) (SSL *ssl, int is_export, int keysize);
-#endif
```
Comment (t8m) [inline ssl/ssl_local.h:2014]: Perhaps keep those ifndefs?

File: ssl/statem/statem_srvr.c [hunk 11]
```diff
@@ -2466,7 +2460,7 @@ int tls_construct_server_key_exchange(SSL *s, WPACKET *pkt)
         } else {
             pkdhp = cert->dh_tmp;
         }
-#if !defined(OPENSSL_NO_DH) && !defined(OPENSSL_NO_DEPRECATED_3_0)
+#if !defined(OPENSSL_NO_DEPRECATED_3_0)
```
Comment (t8m) [inline ssl/statem/statem_srvr.c:2463]: Keep as is here?

File: ssl/tls_depr.c [hunk 12]
```diff
@@ -144,9 +144,9 @@ HMAC_CTX *ssl_hmac_get0_HMAC_CTX(SSL_HMAC *ctx)
 }
 
 /* Some deprecated public APIs pass DH objects */
-# ifndef OPENSSL_NO_DH
 EVP_PKEY *ssl_dh_to_pkey(DH *dh)
 {
+# ifndef OPENSSL_NO_DH
```
Comment (t8m) [inline ssl/tls_depr.c:149]: Changes in this file would not be needed if you revert the changes I proposed above, imo.

**Review by t8m — COMMENTED** [review]
I propose keep the CTRLs that require EC_KEY and DH types to be disabled by NO_EC and NO_DH respectively. Otherwise LGTM.

**Comment by t8m** [comment]
This is linked to evaluated issues.

**Comment by mattcaswell** [comment]
> I propose keep the CTRLs that require EC_KEY and DH types to be disabled by NO_EC and NO_DH respectively. Otherwise LGTM.

The advantage of doing things this way is that it keeps all low level stuff in one file. It's only the tls_depr.c file that has to concern itself with low level APIs at all - rather than littering the guards related to this throughout the rest of the library. So, I prefer the way that it is in this PR to keeping it as it is. From an application perspective the behaviour is the same - if an application calls a ctrl that is not supported it gets a 0 return back.

The `dh_tmp_cb`case  is slightly different although the motivation is the same, i.e. keep all the low level stuff in one file and avoid spreading the guards around the codebase.

**Comment by t8m** [comment]
OK. Approved then. The trivial conflict in CHANGES.md can be resolved during the merge.

**Review by t8m — APPROVED** [review]

**Comment by mattcaswell** [comment]
Pushed. Thanks. I resolved the trivial CHANGES.md conflict during merge.
